appsdk/sdk.go

```diff
@@ -320,223 +312,94 @@ func (sdk *AppFunctionsSDK) Initialize() error {
-	}
-
-	bootTimeout, err := time.ParseDuration(configuration.Service.BootTimeout)
-	if err != nil {
-		fmt.Printf("warning- failed to parse Service.BootTimeout, use the default %s: %v\n",
-			internal.BootTimeoutDefault.String(), err)
-		bootTimeout = internal.BootTimeoutDefault
-	}
-
-	timeElapsed := time.Since(timeStart)
-
-	// Bootstrap retry loop to ensure all dependencies are ready before continuing.
-	until := time.Now().Add(bootTimeout - timeElapsed)
-	for time.Now().Before(until) {
-		if !configurationInitialized {
-			err := sdk.initializeConfiguration(configuration)
-			if err != nil {
-				fmt.Printf("failed to initialize Registry: %v\n", err)
-				goto ContinueWithSleep
-			}
-			configurationInitialized = true
-			fmt.Printf("Configuration & Registry initialized")
-		}
-
-		if !loggerInitialized {
-			loggingTarget, err := sdk.setLoggingTarget()
-			if err != nil {
-				fmt.Printf("logger initialization failed: %v", err)
-				goto ContinueWithSleep
-			}
-
-			sdk.LoggingClient = logger.NewClient(
-				sdk.ServiceKey,
-				sdk.config.Logging.EnableRemote,
-				loggingTarget,
-				sdk.config.Writable.LogLevel,
-			)
-			sdk.LoggingClient.Info("Logger successfully initialized")
-			sdk.edgexClients.LoggingClient = sdk.LoggingClient
-			loggerInitialized = true
-		}
-
-		// Verify that Core Services major version matches this SDK's major version
-		if !sdk.validateVersionMatch() {
-			return fmt.Errorf("core service's version is not compatible with SDK's version")
-		}
-
-		if !secretProviderInitialized {
-			if err := sdk.initializeSecretProvider(); err != nil {
-				return err
-			}
-			secretProviderInitialized = true
+			sdk.ServiceKey = strings.Replace(sdk.ServiceKey, ProfileSuffixPlaceholder, "-"+sdkFlags.Profile(), 1)
 		}
-
-		// Currently only need the database if store and forward is enabled
-		if sdk.config.Writable.StoreAndForward.Enabled {
-			if !databaseInitialized {
-				if sdk.initializeStoreClient() != nil {
-
-					// Error already logged
-					goto ContinueWithSleep
-				}
-
-				databaseInitialized = true
-			}
-		}
-
-		sdk.initializeClients()
-		sdk.LoggingClient.Info("Clients initialized")
-
-		// This is the last dependency so can break out of the retry loop.
-		bootstrapComplete = true
-		break
-
-	ContinueWithSleep:
-		time.Sleep(time.Second * time.Duration(1))
 	}
 
-	if !bootstrapComplete {
-		return fmt.Errorf("bootstrap retry timed out")
-	}
+	sdk.config = &common.ConfigurationStruct{}
+	dic := di.NewContainer(di.ServiceConstructorMap{
+		container.ConfigurationName: func(get di.Get) interface{} {
+			return sdk.config
+		},
+	})
 
+	sdk.appCtx, sdk.appCancelCtx = context.WithCancel(context.Background())
 	sdk.appWg = &sync.WaitGroup{}
 
-	if sdk.useRegistry {
-		sdk.appWg.Add(1)
-		go sdk.listenForConfigChanges()
-	}
+	var deferred bootstrap.Deferred
+	var successful bool
+	var configUpdated config.UpdatedStream = make(chan struct{})
+
+	sdk.appWg, deferred, successful = bootstrap.RunAndReturnWaitGroup(
```

Comment:
i see that it returns a true/false for successful, but errors aren't returned here. In the future i think it would be nice to report any errors up to here and log them out here rather than in each bootstrap func.

Reply:
It is also nice for errors to be logged close to the processing where they occurred since the log contains the line number of the log statement....